nomad/blocked_evals.go

```diff
@@ -0,0 +1,303 @@
+	// blocked evaluations. These should be marked as cancelled since only one
+	// blocked eval is neeeded bper job.
+	duplicates []*structs.Evaluation
+
+	// duplicateCh is used to signal that a duplicate eval was added to the
+	// duplicate set. It can be used to unblock waiting callers looking for
+	// duplicates.
+	duplicateCh chan struct{}
+
+	// stopCh is used to stop any created goroutines.
+	stopCh chan struct{}
+}
+
+// BlockedStats returns all the stats about the blocked eval tracker.
+type BlockedStats struct {
+	// TotalEscaped is the total number of blocked evaluations that have escaped
+	// computed node classes.
+	TotalEscaped int
+
+	// TotalBlocked is the total number of blocked evaluations.
+	TotalBlocked int
+}
+
+// NewBlockedEvals creates a new blocked eval tracker that will enqueue
+// unblocked evals into the passed broker.
+func NewBlockedEvals(evalBroker *EvalBroker) *BlockedEvals {
+	return &BlockedEvals{
+		evalBroker:  evalBroker,
+		captured:    make(map[string]*structs.Evaluation),
+		escaped:     make(map[string]*structs.Evaluation),
+		jobs:        make(map[string]struct{}),
+		unblockCh:   make(chan string, unblockBuffer),
+		duplicateCh: make(chan struct{}),
+		stopCh:      make(chan struct{}),
+		stats:       new(BlockedStats),
+	}
+}
+
+// Enabled is used to check if the broker is enabled.
+func (b *BlockedEvals) Enabled() bool {
+	b.l.RLock()
+	defer b.l.RUnlock()
+	return b.enabled
+}
+
+// SetEnabled is used to control if the broker is enabled. The broker
+// should only be enabled on the active leader.
+func (b *BlockedEvals) SetEnabled(enabled bool) {
+	b.l.Lock()
+	b.enabled = enabled
+	if !b.running {
+		b.running = true
+		go b.unblock()
+	}
+	b.l.Unlock()
+	if !enabled {
+		b.Flush()
+	}
+}
+
+// Block tracks the passed evaluation and enqueues it into the eval broker when
+// a suitable node calls unblock.
+func (b *BlockedEvals) Block(eval *structs.Evaluation) {
+	b.l.Lock()
+	defer b.l.Unlock()
+
+	// Do nothing if not enabled
+	if !b.enabled {
+		return
+	}
+
+	// Check if the job already has a blocked evaluation
+	if _, existing := b.jobs[eval.JobID]; existing {
+		b.duplicates = append(b.duplicates, eval)
+
+		// Unblock any waiter.
+		select {
+		case b.duplicateCh <- struct{}{}:
+		default:
+		}
+
+		return
+	}
+
+	b.stats.TotalBlocked++
+	b.jobs[eval.JobID] = struct{}{}
+	if eval.EscapedComputedClass {
+		b.escaped[eval.ID] = eval
+		b.stats.TotalEscaped++
+		return
+	}
+
+	b.captured[eval.ID] = eval
+}
+
+// Unblock causes any evaluation that could potentially make progress on a
+// capacity change on the passed computed node class to be enqueued into the
+// eval broker.
+func (b *BlockedEvals) Unblock(computedClass string) {
+	// Do nothing if not enabled
+	if !b.enabled {
+		return
+	}
+
+	b.unblockCh <- computedClass
+}
+
+func (b *BlockedEvals) unblock() {
+	for {
+		select {
+		case <-b.stopCh:
+			return
+		case computedClass := <-b.unblockCh:
+			b.l.Lock()
```

Comment:
I would move the inner handling of this into another method to make that method and unblock cleaner, also allows you to use a defer unlock which is much simpler.